fe/fe-core/src/main/java/org/apache/doris/analysis/StmtRewriter.java

```diff
@@ -1365,4 +1370,116 @@ public static boolean rewriteByPolicy(StatementBase statementBase, Analyzer anal
+                aggExprs.add(func);
+                return true;
+            }
+        }
+    }
+
+    public static boolean rewriteForRandomDistribution(StatementBase statementBase, Analyzer analyzer)
+            throws UserException {
+        boolean reAnalyze = false;
+        if (!(statementBase instanceof SelectStmt)) {
+            return false;
+        }
+        SelectStmt selectStmt = (SelectStmt) statementBase;
+        for (int i = 0; i < selectStmt.fromClause.size(); i++) {
+            TableRef tableRef = selectStmt.fromClause.get(i);
+            // Recursively rewrite subquery
+            if (tableRef instanceof InlineViewRef) {
+                InlineViewRef viewRef = (InlineViewRef) tableRef;
+                if (rewriteForRandomDistribution(viewRef.getQueryStmt(), viewRef.getAnalyzer())) {
+                    reAnalyze = true;
+                }
+                continue;
+            }
+            // already has agg and group by info
+            if (selectStmt.hasAggInfo() && selectStmt.hasGroupByClause()) {
+                continue;
+            }
+            TableIf table = tableRef.getTable();
+            if (!(table instanceof OlapTable)) {
+                continue;
+            }
+            OlapTable olapTable = (OlapTable) table;
+            if (olapTable.getKeysType() != KeysType.AGG_KEYS) {
+                continue;
+            }
+            DistributionInfo distributionInfo = olapTable.getDefaultDistributionInfo();
+            if (distributionInfo.getType() != DistributionInfo.DistributionInfoType.RANDOM) {
+                continue;
+            }
+
+            SelectList selectList = selectStmt.getSelectList();
+            SelectList newSelectList = new SelectList();
+            ArrayList<Expr> groupingExprs = new ArrayList<>();
+            ArrayList<FunctionCallExpr> aggExprs = new ArrayList<>();
+            boolean canRewrite = true;
+            for (SelectListItem item : selectList.getItems()) {
+                if (item.isStar()) {
+                    TupleDescriptor desc = tableRef.getDesc();
+                    for (Column col : desc.getTable().getBaseSchema()) {
+                        SlotRef slot = new SlotRef(null, col.getName());
+                        slot.setTable(desc.getTable());
+                        slot.setTupleId(desc.getId());
+                        slot.setDesc(desc.getColumnSlot(col.getName()));
+                        if (!rewriteSelectList(slot, newSelectList, groupingExprs, aggExprs)) {
+                            canRewrite = false;
+                            break;
+                        }
+                    }
+                    if (!canRewrite) {
+                        break;
+                    }
+                } else {
+                    Expr expr = item.getExpr();
+                    // just for SlotRef
+                    if (!(expr instanceof SlotRef)) {
+                        break;
```

Comment:
simply break is not enough, try sql` select citycode, username, pv, siteid + 1 from ${tableName} order by siteid;`

Reply:
I had refactored the logical of rewriting, instead of change original selectStmt, I add an aggregation node for pre-agg, which ignores the complexity of the expression in the query.